synapse/types/__init__.py

```diff
@@ -664,6 +664,12 @@ def __attrs_post_init__(self) -> None:
 
     @classmethod
     async def parse(cls, store: "PurgeEventsStore", string: str) -> "RoomStreamToken":
+        # Check it looks like a Synapse token first. We do this so that
+        # we don't log exceptions for obviously incorrect tokens.
+        if not string or string[0] not in ("s", "t", "m"):
+            logger.warning("Invalid token %r", string)
+            raise SynapseError(400, f"Invalid room stream token {string:!r}")
```

Comment:
Perhaps we should think about just ignoring these obviously wrong tokens altogether or log at a much lower level.

```suggestion
        # Check it looks like a Synapse token first. We do this so that
        # we don't log exceptions for obviously incorrect tokens.
        if not string or string[0] not in ("s", "t", "m"):
            logger.debug("Invalid token %r", string)
            raise SynapseError(400, f"Invalid room stream token {string:!r}")
```

Reply:
Yeah, I was going to suggest at least only logging the empty string case as `debug`. It's plausible that a non-empty token is from a future synapse (after upgrade/downgrade) or something. However, I realised we'll log these tokens anyway when we return the 400 (as we log the error string), so actually adding an extra log line here is pointless, so I've removed it entirely.